enhancements/installer/custom-dns.md

```diff
@@ -0,0 +1,574 @@
+of installer configuration (also called as install-config). [TODO: can we use
+teh same validation method for the Ingress service configuration?]
+
+3. When the custom DNS configuration mentioned in #1 is present, skip generating
+terraform configuration for DNS configuration in Route53.
+
+4. The Ingress Controller configures and manages the DNS configuration for the
+Ingress service. The Ingress Controller needs to be made aware of the presence
+of the custom DNS via the DNS manifest so that it skips configuring the cloud
+DNS for the Ingress service.
+
+5. The Ingress Controller continues managing the Load Balancer configuration for
+the Ingress service.
+
+6. When the custom DNS solution is enabled, installer tries to resolve the
+API Server URL to obtain the IP adderesses within each subnet in the machine
+network and validate them.
+
+7. The API-Int Server would be configured to resolve to the IP addresses provided
+by the user. (It could be configured to resolve to the same IP addresses
+as the API Server.) It is the responsibilty of the Installer to provide these
+IP addresses to the machine-config-operator (MCO) via the Infrastructure CR.
+
+8. The Installer adds the above API-Int Server IPs to the AWSPlatformStatus
+struct within the Infrastructure Status. A new field needs to be added to
+AWSPlatformStatus for this.
+
+9. During the bootstrap phase, it appears that only the API and Ingress services
+need to be resolvable. So, no additional changes are needed during the bootstrap
+phase to support this feature.
+
+10. The MCO is also responsible for generating manifests for running CoreDNS on
+the master and worker nodes using information provided in the Infrastructure CR.
+
+### API Extensions
+
+#### api.dns [https://github.com/openshift/api/blob/master/config/v1/types_dns.go]
+
+We need a way to inform the Ingress Controller to not configure the DNS records
+for the Ingress service in the cloud DNS when the custom DNS has already been
+configured. So, adding a string config to the DNSSpec of the DNS CR to
+indicate whether the DNS configuration has already been handled externally
+should take care of that.
+
+```go
+
+type DNSSpec struct {
+// <snip>
+
+// ingressDNSType indicates whether the DNS configuration for the Ingress
+// service has been handled externally by a custom DNS solution.
+// A value of "External" indicates that the Ingress Controller does not have
+// to configure the cloud DNS and a default value of "Internal" indicates
+// that the Ingress Controller is responsible for the DNS configuration.
+
+// Once set, this field cannot be changed.
+// +optional
+IngressDNSType string `json:"ingressDNSType"`
+}
+```
+
+#### Infrastructure CR [https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go]
+
+```go
+
+type AWSPlatformStatus struct {
+// <snip>
+
+// A field with the name APIServerInternalIPs is used in the BareMetal platform to
```

Comment:
If we have another field with a similar purpose elsewhere, should we use the same name? Would it make sense to move that field to a section of the API that is not platform-specific?

Reply:
Yeah, this is confusing. 
APIServerInternalIPs[] is used in the baremetal platform to hold the API-Int server IP. In the dual-stack case, we can expect this config to have 2 IP addresses, 1 for each protocol. In IPv4 or IPv6 deployments, this config is expected to hold only 1 IP. 

In this custom DNS feature, we expect the user to provide IP addresses for the API-Int server when custom DNS feature is enabled on public cloud platforms. We expect one IP address per subnet.

Although both config items hold IP Addresses that the API-Int server need to be configured with, the semantics seem different based on the platform. So, I chose a different name so as not cause confusion for the customer.

Giving them the same name and moving it to a platform independent section of install-config definitely seems cleaner but I am worried about regressions.